Add an example for building a flex template for a python pipeline with extra dependencies and a custom container. (#11252)
GoogleCloudPlatform/python-docs-samples: dataflow/flex-templates/pipeline_with_dependencies/README.md

# Dataflow flex template: a pipeline with dependencies and a custom container image.

This project is created to illustrate the following setup of a Dataflow Python pipeline:
- The pipeline is a package that consists of [multiple files](https://beam.apache.org/documentation/sdks/python-pipeline-dependencies/#multiple-file-dependencies).
- The pipeline has at least one depenendency that is not provided in the default Dataflow runtime environment.
- The workflow uses a [custom container image](https://cloud.google.com/dataflow/docs/guides/using-custom-containers) to preinstall dependencies and define the pipeline runtime environment.
- The workflow uses a [Dataflow Flex template](https://cloud.google.com/dataflow/docs/concepts/dataflow-templates) to control the pipeline submission environment.
- The runtime and submission environment use same set of Python dependencies and can be created in a reproducible manner.

> To illustrate this setup, we will use a pipepline that finds a longest word in an input file, crates a [FIGLet text banner](https://en.wikipedia.org/wiki/FIGlet) from of it using [pyfiglet](https://pypi.org/project/pyfiglet/), and outputs it in another file.

Reviewer: ```suggestion
To illustrate this setup, we use a pipeline that does the following:

1. Finds the longest word in an input file
2. Creates a [FIGLet text banner](https://en.wikipedia.org/wiki/FIGlet) from of it using [pyfiglet](https://pypi.org/project/pyfiglet/)
3. Outputs the text banner in another file
```

For the last step, I changed "it" to "text banner," but please verify that that's correct.

## The structure of the example

> The `my_package` directory together with `setup.py` comprise the pipeline package. The package defines the pipeline, its dependenicies and input parameters. It is possible to define multiple pipelines in the same package. The `my_package.launcher` is used to submit the pipeline to a runner.

Reviewer: ```suggestion
The pipeline package is comprised of the `my_package` directory and the `setup.py` file. The package defines the pipeline, the pipeline dependencies, and the input parameters. You can define multiple pipelines in the same package. The `my_package.launcher` file is used to submit the pipeline to a runner.
```

The `main.py` file provides a top-level entrypoint to trigger the pipeline launcher from a
launch environment.

> The `Dockerfile` defines the runtime environment for our pipeline, and also configures the Flex Template, so that we can reuse runtime image for building our Flex Template.

Reviewer: ```suggestion
The `Dockerfile` defines the runtime environment for the pipeline. It also configures the Flex Template, which lets you reuse the runtime image to build the Flex Template.
```

The`Dockerfile.flex_template` and `flex_template_cloudbuild.yaml` files provide the remaining instructions to build the Flex template.

> The `requirements.txt` defines all Python packages in the dependency chain of our pipeline package. We use it to create reproducible Python environments in our Docker images.

Reviewer: ```suggestion
The `requirements.txt` file defines all Python packages in the dependency chain of the pipeline package. Use it to create reproducible Python environments in the Docker images.
```

## Before you begin

Make sure you have followed the
[Dataflow setup instructions](../../README.md).

## Create a Cloud Storage bucket

```sh
export PROJECT="project-id"
export BUCKET="your-bucket"
export REGION="us-central1"
gsutil mb -p $PROJECT gs://$BUCKET
```

## Create an Artifact Registry repository

```sh
export REPOSITORY="your-repository"
gcloud artifacts repositories create $REPOSITORY \
    --repository-format=docker \
    --location=$REGION

gcloud artifacts repositories create $REPOSITORY \
    --repository-format=docker \
    --location=$REGION \
    --project $PROJECT
```

> ## Build a Docker image for pipeline runtime environment

Reviewer: ```suggestion
## Build a Docker image for the pipeline runtime environment
```

Using a [custom SDK container image](https://cloud.google.com/dataflow/docs/guides/using-custom-containers)
allows flexible customizations of the runtime environment.

> We use the custom container image to preinstall all dependencies of our pipeline ahead of job submission and to have a reproducible runtime environment.

Reviewer: ```suggestion
We use the custom container image both to preinstall all of the pipeline dependencies before job submission and to create a reproducible runtime environment.
```

To illustrate customizations, we use a [custom base base image](https://cloud.google.com/dataflow/docs/guides/build-container-image#use_a_custom_base_image) to build the SDK container image.

```sh
# Use a unique tag to version artifacts we build.
export TAG=`date +%Y%m%d-%H%M%S`
export SDK_CONTAINER_IMAGE="$REGION-docker.pkg.dev/$PROJECT/$REPOSITORY/my_base_image:$TAG"

gcloud builds submit .  --tag $SDK_CONTAINER_IMAGE --project $PROJECT
```

You must rebuild the base image when your pipeline dependencies change.

> ## Build the Flex template

Reviewer: ```suggestion
## Build the Flex Template
```

We will use the [runtime image as the base image](https://cloud.google.com/dataflow/docs/guides/templates/configuring-flex-templates#use_custom_container_images)

>  when building our Flex templates.

Reviewer: ```suggestion
```

This allows us to reduce the number of Docker images to maintain and

> it ensures that the pipeline  uses the same dependencies at submission and at runtime.

Reviewer: ```suggestion
It also ensures that the pipeline uses the same dependencies at submission and at runtime.
```

```sh
export TEMPLATE_FILE=gs://$BUCKET/longest-word-$TAG.json
export TEMPLATE_IMAGE=$REGION-docker.pkg.dev/$PROJECT/$REPOSITORY/my_template_image:$TAG
```

```sh
gcloud dataflow flex-template build $TEMPLATE_FILE  \
    --image-gcr-path $TEMPLATE_IMAGE \
    --sdk-language "PYTHON" \
    --flex-template-base-image $SDK_CONTAINER_IMAGE \
    --py-path "."  \
    --env "FLEX_TEMPLATE_PYTHON_PY_FILE=main.py" \
    --env "FLEX_TEMPLATE_PYTHON_SETUP_FILE=setup.py" \
    --project $PROJECT
```

> Note: Older versions of gcloud might not support AR and require using GCR registry images in the `--flex-template-base-image` option.

Author: I'll have a better wording to recommend the option below or upgrade gcloud.

This project is created to illustrate the following setup of a Dataflow Python pipeline:
- The pipeline is a package that consists of [multiple files](https://beam.apache.org/documentation/sdks/python-pipeline-dependencies/#multiple-file-dependencies).
- The pipeline has at least one dependency that is not provided in the default Dataflow runtime environment.
- The workflow uses a [custom container image](https://cloud.google.com/dataflow/docs/guides/using-custom-containers) to preinstall dependencies and to define the pipeline runtime environment.
- The workflow uses a [Dataflow Flex Template](https://cloud.google.com/dataflow/docs/concepts/dataflow-templates) to control the pipeline submission environment.
- The runtime and submission environment use same set of Python dependencies and can be created in a reproducible manner.

To illustrate this setup, we use a pipeline that does the following:

1. Finds the longest word in an input file
2. Creates a [FIGLet text banner](https://en.wikipedia.org/wiki/FIGlet) from of it using [pyfiglet](https://pypi.org/project/pyfiglet/)
3. Outputs the text banner in another file

## The structure of the example

> The pipeline package is comprised of the `my_package` directory and the `setup.py` file. The package defines the pipeline, the pipeline dependencies, and the input parameters. You can define multiple pipelines in the same package. The `my_package.launcher` module is used to submit the pipeline to a runner.

Reviewer 2: Maybe just my preference: I feel like the directory structure is easier to follow (e.g., https://github.com/google/dataflow-ml-starter/tree/main?tab=readme-ov-file#directory-structure)

Author: Yes, I like your structure and it seems to me that this example follows almost same structure. the only notable difference I see is that you fold the entrypoint inside the launcher: https://github.com/google/dataflow-ml-starter/blob/2c591d3d274f1ef9f2f79db026093cbf2b7d36d5/my_project/run.py#L97

I also considered that and went back-and-forth a few times. I finally chose to keep entrypoint outside 
1) for consistency with beam docs and juliaset example: https://beam.apache.org/documentation/sdks/python-pipeline-dependencies/#multiple-file-dependencies (recently updated)
2) some users supply their pipeline in a package, but pass a package tarball via --extra_package and don't use --setup_file option. in such setup, source of the package doesn't need to in the launch environment, as long as the package is installed, however to launch the pipeline from flex template, one needs to provide a py file with the entrypoint: FLEX_TEMPLATE_PYTHON_PY_FILE=/path/to/launcher.py . If that top-level entrypoint lives separately from the package it might be easier to accomodate both setups.

Do you see any other significant differences?

Reviewer 2: I noticed this but I think it is fine. The ML starter one tries to keep all the python code in one place. Your way is also fine. :)

The `main.py` file provides a top-level entrypoint to trigger the pipeline launcher from a
launch environment.

The `Dockerfile` defines the runtime environment for the pipeline. It also configures the Flex Template, which lets you reuse the runtime image to build the Flex Template.

The`Dockerfile.flex_template` and `flex_template_cloudbuild.yaml` files provide the remaining instructions to build the Flex Template.

The `requirements.txt` file defines all Python packages in the dependency chain of the pipeline package. Use it to create reproducible Python environments in the Docker images.

## Before you begin

Follow the
[Dataflow setup instructions](../../README.md).

## Create a Cloud Storage bucket

```sh
export PROJECT="project-id"
export BUCKET="your-bucket"
export REGION="us-central1"
gsutil mb -p $PROJECT gs://$BUCKET
```

## Create an Artifact Registry repository

```sh
export REPOSITORY="your-repository"
gcloud artifacts repositories create $REPOSITORY \
    --repository-format=docker \
    --location=$REGION

gcloud artifacts repositories create $REPOSITORY \
    --repository-format=docker \
    --location=$REGION \

>     --project $PROJECT

Author: ```suggestion
    --project $PROJECT
    
gcloud auth configure-docker $REGION-docker.pkg.dev
```

> ```

Reviewer 2: I remember I have to do these as well to get the access:
```
gcloud auth login
gcloud auth application-default login
# replace it with the appropriate region
gcloud auth configure-docker us-docker.pkg.dev
# or if you use docker-credential-gcr
docker-credential-gcr configure-docker --registries=us-docker.pkg.dev
```
Maybe it is different if using `gcloud builds`?

Author: > gcloud auth application-default login

This is mentioned in Dataflow prerequisite instructions.

> gcloud auth configure-docker us-docker.pkg.dev

Will add this. It's easy, and it helps if you use local docker.

export TAG=`date +%Y%m%d-%H%M%S`
export SDK_CONTAINER_IMAGE="$REGION-docker.pkg.dev/$PROJECT/$REPOSITORY/my_base_image:$TAG"

gcloud builds submit .  --tag $SDK_CONTAINER_IMAGE --project $PROJECT
```

You must rebuild the base image when your pipeline dependencies change.

## Build the Flex Template

To build the Flex Templates, use the [runtime image as the base image](https://cloud.google.com/dataflow/docs/guides/templates/configuring-flex-templates#use_custom_container_images).
Using the runtime image as the base image allows us to reduce the number of Docker images that need to be maintained.
It also ensures that the pipeline uses the same dependencies at submission and at runtime.

```sh
export TEMPLATE_FILE=gs://$BUCKET/longest-word-$TAG.json
export TEMPLATE_IMAGE=$REGION-docker.pkg.dev/$PROJECT/$REPOSITORY/my_template_image:$TAG
```

```sh
gcloud dataflow flex-template build $TEMPLATE_FILE  \
    --image-gcr-path $TEMPLATE_IMAGE \
    --sdk-language "PYTHON" \
    --flex-template-base-image $SDK_CONTAINER_IMAGE \
    --py-path "."  \
    --env "FLEX_TEMPLATE_PYTHON_PY_FILE=main.py" \
    --env "FLEX_TEMPLATE_PYTHON_SETUP_FILE=setup.py" \
    --project $PROJECT
```

> Note: Older versions of gcloud might not support AR and require using GCR registry images in the `--flex-template-base-image` option.

Author: ```suggestion
Note: Older versions of `gcloud` might not support Artifact Registry in the `--flex-template-base-image` option. You might need to upgrade `gcloud` or build the Flex Template from a `Dockerfile` using the instructions below.
```

If you prefer to have a separate file to configure and build the Flex Template image (included in the example), such as a `Dockefile.flex_template` file, you can build the Flex Template as follows:

```sh
gcloud builds submit --config flex_template_cloudbuild.yaml \
    --substitutions="_TEMPLATE_IMAGE=${TEMPLATE_IMAGE},_BASE_IMAGE=${SDK_CONTAINER_IMAGE}" \
    --project $PROJECT

gcloud dataflow flex-template build $TEMPLATE_FILE  \
    --image $TEMPLATE_IMAGE \
    --sdk-language "PYTHON" \
    --project $PROJECT
```

> ## Run the template

Reviewer 2: I suggest we could add a new optional section to check the container to verify the installed packages.

    --template-file-gcs-location $TEMPLATE_FILE \
    --region $REGION \
    --staging-location "gs://$BUCKET/staging" \
    --parameters input="gs://dataflow-samples/shakespeare/hamlet.txt" \
    --parameters output="gs://$BUCKET/output" \
    --parameters sdk_container_image=$SDK_CONTAINER_IMAGE \
    --project $PROJECT
```

After the pipeline finishes, use the following command to inspect the output:
```
gsutil cat gs://$BUCKET/output*
```

## Optional: Update the dependencies in requirements.txt and rebuild the Docker images

The top-level pipeline dependencies are defined in the `install_requires` section of the `setup.py` file.

The `requirements.txt` file pins all Python dependencies, including transitive dependencies, that must be installed in the Docker base images. This step produces a reproducible set of dependencies every time the image is built.
Version control the `requirements.txt` file together with the rest of pipeline code.

When the dependencies of your pipeline change or when you want to use the latest available versions of packages in the pipeline's dependency chain,  regenerate the `requirements.txt` file:

```

>     python3.11 -m pip install pip-tools   # Use a consistent minor version of Python throughout the project.

Reviewer 2: Do we need to pin python to 3.11? shall we just do `python3`?

Author: pinning is better since dependency resolution can differ between minor versions. it can also be sensitive to which OS is being used...

Author: example: numpy is no longer released on some earlier versions of python, so it will resolve to an earlier version.

# Dataflow flex template: a pipeline with dependencies and a custom container image.

[![Open in Cloud Shell](http://gstatic.com/cloudssh/images/open-btn.svg)](https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=dataflow/flex-templates/streaming_beam/README.md)

> This project is created to illustrate the following setup of a Dataflow Python pipeline:

Reviewer: ```suggestion
This project illustrates the following Dataflow Python pipeline setup:
```

1. [Enable the Cloud Build API](https://console.cloud.google.com/flows/enableapi?apiid=cloudbuild.googleapis.com).

1. Clone the [`python-docs-samples` repository](https://github.com/GoogleCloudPlatform/python-docs-samples)
and navigate to the code sample.

    ```sh
    git clone https://github.com/GoogleCloudPlatform/python-docs-samples.git
    cd python-docs-samples/dataflow/flex-templates/streaming_beam
    ```

## Create a Cloud Storage bucket

```sh
export PROJECT="project-id"
export BUCKET="your-bucket"
export REGION="us-central1"
gsutil mb -p $PROJECT gs://$BUCKET
```

## Create an Artifact Registry repository

```sh
export REPOSITORY="your-repository"
gcloud artifacts repositories create $REPOSITORY \
    --repository-format=docker \
    --location=$REGION

gcloud artifacts repositories create $REPOSITORY \
    --repository-format=docker \
    --location=$REGION \
    --project $PROJECT

gcloud auth configure-docker $REGION-docker.pkg.dev
```

## Build a Docker image for the pipeline runtime environment

Using a [custom SDK container image](https://cloud.google.com/dataflow/docs/guides/using-custom-containers)
allows flexible customizations of the runtime environment.

> We use the custom container image both to preinstall all of the pipeline dependencies before job submission and to create a reproducible runtime environment.

Reviewer: ```suggestion
This example uses the custom container image both to preinstall all of the pipeline dependencies before job submission and to create a reproducible runtime environment.
```

> To illustrate customizations, we use a [custom base base image](https://cloud.google.com/dataflow/docs/guides/build-container-image#use_a_custom_base_image) to build the SDK container image.

Reviewer: ```suggestion
To illustrate customizations, a [custom base base image](https://cloud.google.com/dataflow/docs/guides/build-container-image#use_a_custom_base_image) is used to build the SDK container image.
```

> We include the Flex Template launcher in the SDK container image, which makes it possible to [use the SDK container image to build a Flex Template](https://cloud.google.com/dataflow/docs/guides/templates/configuring-flex-templates#use_custom_container_images).

Reviewer: ```suggestion
The Flex Template launcher is included in the SDK container image, which makes it possible to [use the SDK container image to build a Flex Template](https://cloud.google.com/dataflow/docs/guides/templates/configuring-flex-templates#use_custom_container_images).
```

```sh
# Use a unique tag to version artifacts we build.
export TAG=`date +%Y%m%d-%H%M%S`
export SDK_CONTAINER_IMAGE="$REGION-docker.pkg.dev/$PROJECT/$REPOSITORY/my_base_image:$TAG"

gcloud builds submit .  --tag $SDK_CONTAINER_IMAGE --project $PROJECT
```

## Optional: Inspect the Docker image

> If you have a local installation of Docker, you can inspect the image and run the pipeline on the Direct Runner as follows:

Reviewer: ```suggestion
If you have a local installation of Docker, you can inspect the image and run the pipeline by using the Direct Runner:
```

```
docker run --rm -it --entrypoint=/bin/bash $SDK_CONTAINER_IMAGE

# Once the container is created, run:
pip list
python main.py --input requirements.txt --output=/tmp/output
cat /tmp/output*
```

## Build the Flex Template

> We build the Flex Template [from the SDK container image](https://cloud.google.com/dataflow/docs/guides/templates/configuring-flex-templates#use_custom_container_images).

Reviewer: ```suggestion
Build the Flex Template [from the SDK container image](https://cloud.google.com/dataflow/docs/guides/templates/configuring-flex-templates#use_custom_container_images).
```

> Using the runtime image as the Flex Template image allows us to reduce the number of Docker images that need to be maintained.

Reviewer: ```suggestion
Using the runtime image as the Flex Template image reduces the number of Docker images that need to be maintained.
```

It also ensures that the pipeline uses the same dependencies at submission and at runtime.

```sh
export TEMPLATE_FILE=gs://$BUCKET/longest-word-$TAG.json
export TEMPLATE_IMAGE=$REGION-docker.pkg.dev/$PROJECT/$REPOSITORY/my_template_image:$TAG
```

```sh
gcloud dataflow flex-template build $TEMPLATE_FILE  \
    --image $SDK_CONTAINER_IMAGE \
    --sdk-language "PYTHON" \
    --project $PROJECT
```

## Run the template

```sh
gcloud dataflow flex-template run "flex-`date +%Y%m%d-%H%M%S`" \
    --template-file-gcs-location $TEMPLATE_FILE \
    --region $REGION \
    --staging-location "gs://$BUCKET/staging" \
    --parameters input="gs://dataflow-samples/shakespeare/hamlet.txt" \
    --parameters output="gs://$BUCKET/output" \
    --parameters sdk_container_image=$SDK_CONTAINER_IMAGE \
    --project $PROJECT
```

After the pipeline finishes, use the following command to inspect the output:
```
gsutil cat gs://$BUCKET/output*
```

> ## Optional: Update the dependencies in requirements.txt and rebuild the Docker images

Reviewer: ```suggestion
## Optional: Update the dependencies in the requirements file and rebuild the Docker images
```

The top-level pipeline dependencies are defined in the `install_requires` section of the `setup.py` file.

> The `requirements.txt` file pins all Python dependencies, including transitive dependencies, that must be installed in the Docker base images. This step produces a reproducible set of dependencies every time the image is built.

Reviewer: ```suggestion
The `requirements.txt` file pins all Python dependencies, that must be installed in the Docker base images, including the transitive dependencies. This step produces a reproducible set of dependencies every time the image is built.
```

Version control the `requirements.txt` file together with the rest of pipeline code.

> When the dependencies of your pipeline change or when you want to use the latest available versions of packages in the pipeline's dependency chain,  regenerate the `requirements.txt` file:

Reviewer: ```suggestion
When the dependencies of your pipeline change or when you want to use the latest available versions of packages in the pipeline's dependency chain, regenerate the `requirements.txt` file:
```

```
    python3.11 -m pip install pip-tools   # Use a consistent minor version of Python throughout the project.
    pip-compile ./setup.py
```

To reduce the image size and to give preference to the versions already installed in the Apache Beam base image, use a constraints file:

```
   wget https://raw.githubusercontent.com/apache/beam/release-2.54.0/sdks/python/container/py311/base_image_requirements.txt
   pip-compile --constraint=base_image_requirements.txt ./setup.py
```

> Alternatively, use an empty `requirements.txt` file,  build the SDK container Docker image from the Docker file,

Reviewer: ```suggestion
Alternatively, take the following steps:

1. Use an empty `requirements.txt` file.
1. Build the SDK container Docker image from the Docker file.
```

> collect the output of `pip freeze` at the last stage of the Docker build, and seed the `requirements.txt` with that content.

Reviewer: ```suggestion
1. Collect the output of `pip freeze` at the last stage of the Docker build.
1. Seed the `requirements.txt` file with that content.
```

export REPOSITORY="your-repository"
gcloud artifacts repositories create $REPOSITORY \
    --repository-format=docker \
    --location=$REGION

gcloud artifacts repositories create $REPOSITORY \
    --repository-format=docker \
    --location=$REGION \
    --project $PROJECT

gcloud auth configure-docker $REGION-docker.pkg.dev
```

## Build a Docker image for the pipeline runtime environment

Using a [custom SDK container image](https://cloud.google.com/dataflow/docs/guides/using-custom-containers)
allows flexible customizations of the runtime environment.

This example uses the custom container image both to preinstall all of the pipeline dependencies before job submission and to create a reproducible runtime environment.

To illustrate customizations, a [custom base base image](https://cloud.google.com/dataflow/docs/guides/build-container-image#use_a_custom_base_image) is used to build the SDK container image.

The Flex Template launcher is included in the SDK container image, which makes it possible to [use the SDK container image to build a Flex Template](https://cloud.google.com/dataflow/docs/guides/templates/configuring-flex-templates#use_custom_container_images).

```sh

> # Use a unique tag to version artifacts we build.

Author: ```suggestion
# Use a unique tag to version the artifacts that are built.
```
